### PR TITLE
Add support for MariaDB

### LINKED ISSUE #8904: Self-hosted MariaDB support
### What

Teleport database access should support self-hosted MariaDB databases.

### How

MariaDB is fully compatible with MySQL protocol (to the point where `mysql` and `mariadb` CLI clients can be used interchangeably) which database access already supports. So chances are, MariaDB is already supported, we just need to validate it and write a documentation guide.

Action items:

- [ ] Use X.509 authentication with MariaDB similar to MySQL.
- [ ] Update MySQL integration to use `mariadb` shell client if it's available. Many guides recommend using this client even for MySQL databases.
- [ ] Write documentation guide.
  - If it's similar to MySQL (for instance, only differs in the way you setup mTLS in the database config), update existing MySQL [guide](https://goteleport.com/docs/database-access/guides/mysql-self-hosted/) - use tabs to show configuration for MySQL vs MariaDB.
  - Otherwise, create a separate guide page.

### PR DESCRIPTION
**Pull request by jakule**
This PR add support for MariaDB. Binary protocol is the same as MySQL, so there is no changes in the backend logic. Default client for MySQL and MariaDB from now on will be `mariadb` or `mysql` if the first one is not available in the system.
I also added tests for all CLI tools run by `tsh db connect`.

closes #8904

### CODE REVIEW

File: tool/tsh/db.go [hunk 1]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+}
+
+// isMySQLBinMariaDBFlavor checks if mysql binary comes from Oracle or MariaDB.
+// true is returned when binary comes from MariaDB, false when from Oracle.
+func (c *cliCommandBuilder) isMySQLBinMariaDBFlavor() (bool, error) {
+	// Check if mysql comes from Oracle or MariaDB
+	mysqlVer, err := c.exe.RunCommand(mysqlBin, "--version")
+	if err != nil {
+		// Looks like incorrect mysql installation or mysql binary is missing.
+		// Assume the Oracle's version and return the command. Nest time when
```
Comment (gabrielcorado) [inline tool/tsh/db.go:693]:
```suggestion
		// Assume the Oracle's version and return the command. Next time when
```

**Review by gabrielcorado — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

File: docs/pages/database-access/guides/rds.mdx [hunk 2]
```diff
@@ -374,7 +374,7 @@ $ tsh db connect postgres-rds
 ```
 
 <Admonition type="note" title="Note">
-  The appropriate database command-line client (`psql`, `mysql`) should be
+  The appropriate database command-line client (`psql`, `mysql`,`mariadb`) should be
```
Comment (r0mant) [inline docs/pages/database-access/guides/rds.mdx:377]:
```suggestion
  The appropriate database command-line client (`psql`, `mysql`, `mariadb`) should be
```

File: tool/tsh/db.go [hunk 3]
```diff
@@ -505,7 +506,43 @@ func WithLocalProxy(host string, port int, caPath string) ConnectCommandFunc {
...
+func (s systemExecer) RunCommand(name string, arg ...string) ([]byte, error) {
+	return exec.Command(name, arg...).Output()
+}
+
+// LookPath is a wrapper for exec.LookPath(...)
+func (s systemExecer) LookPath(file string) (string, error) {
+	return exec.LookPath(file)
+}
+
+type cliCommandBuilder struct {
```
Comment (r0mant) [inline tool/tsh/db.go:532]: What do you think about moving all command-building-related stuff (builder, execer, constants, etc.) into a separate file like `dbcmd.go` or `dbexec.go`?

File: tool/tsh/db.go [hunk 4]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
-		if options.localProxyHost == "localhost" {
+		if c.options.localProxyHost == "localhost" {
 			args = append(args, "--protocol", "TCP")
 		}
 	}
 
+	return args
+}
+
+func (c *cliCommandBuilder) getMariadbArgs() []string {
```
Comment (r0mant) [inline tool/tsh/db.go:627]:
```suggestion
func (c *cliCommandBuilder) getMariaDBArgs() []string {
```

File: tool/tsh/db.go [hunk 5]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
 	return exec.Command(cockroachBin, "sql", "--url",
-		postgres.GetConnString(dbprofile.New(tc, *db, *profile, host, port)))
+		postgres.GetConnString(dbprofile.New(c.tc, *c.db, *c.profile, c.host, c.port)))
 }
 
-func getMySQLCommand(tc *client.TeleportClient, profile *client.ProfileStatus, db *tlsca.RouteToDatabase, options connectionCommandOpts) *exec.Cmd {
-	args := []string{fmt.Sprintf("--defaults-group-suffix=_%v-%v", tc.SiteName, db.ServiceName)}
-	if db.Username != "" {
-		args = append(args, "--user", db.Username)
+func (c *cliCommandBuilder) getMySQLCommonCmdOpts() []string {
```
Comment (r0mant) [inline tool/tsh/db.go:605]: Nit: Explain with a comment what "common" options are (w/r/t MySQL and Maria).

File: tool/tsh/db.go [hunk 6]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+
 	return exec.Command(mysqlBin, args...)
 }
 
-func getMongoCommand(tc *client.TeleportClient, profile *client.ProfileStatus, db *tlsca.RouteToDatabase, host string, port int, options connectionCommandOpts) *exec.Cmd {
+// getMySQLCommand returns mariadb command if the binary is on the path. Otherwise,
+// mysql command is returned. Both mysql versions (MariaDB and Oracle) are supported.
+func (c *cliCommandBuilder) getMySQLCommand() *exec.Cmd {
+	// Check if mariadb client is available. Prefer it over mysql client even if connecting to MySQL server.
+	if c.isMariadbBinAvailable() {
```
Comment (r0mant) [inline tool/tsh/db.go:662]:
```suggestion
	if c.isMariaDBBinAvailable() {
```

File: tool/tsh/db.go [hunk 7]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+func (c *cliCommandBuilder) isMariadbBinAvailable() bool {
+	_, err := c.exe.LookPath(mariadbBin)
+	return err == nil
+}
+
+// isMySQLBinMariaDBFlavor checks if mysql binary comes from Oracle or MariaDB.
+// true is returned when binary comes from MariaDB, false when from Oracle.
+func (c *cliCommandBuilder) isMySQLBinMariaDBFlavor() (bool, error) {
+	// Check if mysql comes from Oracle or MariaDB
+	mysqlVer, err := c.exe.RunCommand(mysqlBin, "--version")
```
Comment (r0mant) [inline tool/tsh/db.go:690]: Wouldn't it be a bit more graceful to check if `mysql` binary exists first and return an appropriate error, like with `mariadb`?

**Review by r0mant — APPROVED** [review]
lgtm with a few nits

File: tool/tsh/db.go [hunk 8]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+	if err != nil {
+		// Looks like incorrect mysql installation or mysql binary is missing.
```
[the remark this answers is not in the document]
Comment (jakule) [reply tool/tsh/db.go:692]: As @r0mant suggested I changed the logic a bit. I changed the comment as now we have an explicit check for `mysql`.

File: tool/tsh/db.go (hunk 1 above)
Comment (jakule) [reply tool/tsh/db.go:693]: Fixed.

File: tool/tsh/db.go [hunk 9]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+	args := c.getMySQLCommonCmdOpts()
+	sslCertPath := c.profile.DatabaseCertPathForCluster(c.tc.SiteName, c.db.ServiceName)
+
+	args = append(args, []string{"--ssl-key", c.profile.KeyPath()}...)
+	args = append(args, []string{"--ssl-ca", c.profile.CACertPath()}...)
+	args = append(args, []string{"--ssl-cert", sslCertPath}...)
+
+	// Flag below verifies "Common Name" check on the certificate provided by the server.
+	if !c.tc.InsecureSkipVerify {
+		args = append(args, "--ssl-verify-server-cert")
```
[the remark this answers is not in the document]
Comment (jakule) [reply tool/tsh/db.go:637]: yes, ref: https://mariadb.com/docs/reference/mdb/cli/mariadb/ssl-verify-server-cert/

File: tool/tsh/db.go (hunk 7 above)
Comment (jakule) [reply tool/tsh/db.go:690]: Added

File: tool/tsh/db.go [hunk 10]
```diff
@@ -518,77 +555,171 @@ func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.P
...
+	}
+
+	return args
+}
+
+func (c *cliCommandBuilder) getMySQLOracleCommand() *exec.Cmd {
```
[the remark this answers is not in the document]
Comment (jakule) [reply tool/tsh/db.go:691]: Added.

File: tool/tsh/db.go (hunk 3 above)
Comment (jakule) [reply tool/tsh/db.go:574]: Moved into `dbcmd.go`.

**Comment by jakule** [comment]
@klizhentas Can you take a look on the docs changes? Thanks.

**Pull request merged**